### DISCUSSION TITLE
Questions concerning NRules.Language

### DISCUSSION THREAD

**Discussion by driekus77** [NRules/NRules#368] [General]
Nice libs both NRules and NRules.Language! Thanks.

But I have some questions concerning NRules.Language:

1.  Is this project still alive? I ask this because I see old issues not beeing handled and last commit is around a year ago.
2.  Who is the owner of the Rule# language? Is it this lib or is the grammer (https://github.com/NRules/NRules.Language/blob/develop/src/NRules.RuleSharp/NRules.RuleSharp/Grammar/RuleSharpParser.g4) a copy from somewhere else?
3.  Can I help in any way?

Thanks again!

**Comment by snikolayev** [comment]
@driekus77 thanks for your interest in NRules. Yes, the project is still alive. NRules.Language is fully functional but supports a subset of all NRules capabilities as compared to the fluent DSL.
Rule# language is defined by the grammar you referenced and is implemented by the NRules.Language library. NRules is the runtime for the engine, and NRules.Language is the Rule# implementation on top of that.
The grammar for Rule# is an amalgamation of the general C# grammar and rules-related meta constructs that I added. There is a reference to the source C# grammar in the Rule# grammar file, from which it was derived.
You are certainly more than welcome to contribute to this project. The idea has always been to evolve Rule# based on the community interest. Most immediately, just supporting all NRules features in Rule# would be very helpful. If you want to contribute, we can talk more on the specifics of what is needed.

**Reply by driekus77** [reply]
Thanks for quick reply!

The first that I miss is support for a statement like:

```csharp
using QTerra.Platform.Application.Projects.EnvironmentalScan.Models;

rule "Test Rule 001"
when
    var inputModel = EnvironmentalScanModel(x => x.MyIDs.Value.Any(y => y>7));
    var resultModel = EnvironmentalScanSubmitResponseModel();
    
then
    resultModel.ResultMessage = "OK";
    Console.WriteLine("CHECK!");
```
So:

1.  Nested navigation into property: x.MyIDs.Value
2.  Nullable(s)? Seems like not supported??
3.  Function calls: ...Any(y => y>7)....

I will fork and dive into the code but maybe you can already give me a few hinds to start-up ;-) ?

**Reply by snikolayev** [reply]
Nested properties and Nullable should already work, I just pushed tests to verify.
`Any(y => ...)` is more complicated. This is a generic extension method with a nested lambda expression. All expressions parsing is happening in https://github.com/NRules/NRules.Language/blob/develop/src/NRules.RuleSharp/NRules.RuleSharp/ExpressionParser.cs
Its job is to produce a System.Linq.Expressions.LambdaExpression, by implementing corresponding visit methods in the ANTLR visitor.
If you want to try tackling this enhancement, I suggest you start by creating a new test under RuleCompilerTests, and see where you hit the unimplemented visit method. I would handle it separately - first, nested lambda expressions, then generic extension methods.
